crates/transpiler/src/passes/high_level_synthesis.rs

```diff
@@ -68,109 +68,172 @@ struct QubitTracker {
-            .filter(|q| !self.ignored[*q] && self.enabled[*q] && self.state[*q]);
+            .filter(|q| !self.ignored[*q] && self.enabled[*q] && self.state[*q])
+            .map(Qubit::new);
         let dirty_ancillas = (0..self.num_qubits)
-            .filter(|q| !self.ignored[*q] && self.enabled[*q] && !self.state[*q]);
-        let out: Vec<usize> = clean_ancillas
+            .filter(|q| !self.ignored[*q] && self.enabled[*q] && !self.state[*q])
+            .map(Qubit::new);
+        let out: Vec<Qubit> = clean_ancillas
             .chain(dirty_ancillas)
             .take(num_qubits)
             .collect();
 
-        for q in &ignored_qubits {
-            self.ignored[*q] = false;
+        for q in ignored_qubits {
+            self.ignored[q.index()] = false;
         }
         out
     }
 
+    /// Replaces the state of the given qubits by their state in the `other` tracker
+    fn replace_state<T: Iterator<Item = Qubit>>(&mut self, other: &QubitTracker, qubits: T) {
+        for q in qubits {
+            self.state[q.index()] = other.state[q.index()]
+        }
+    }
+}
+
+#[pymethods]
+impl QubitTracker {
+    #[new]
+    fn new(num_qubits: usize, qubits_initially_zero: bool) -> Self {
+        QubitTracker {
+            num_qubits,
+            state: vec![qubits_initially_zero; num_qubits],
+            enabled: vec![true; num_qubits],
+            ignored: vec![false; num_qubits],
+        }
+    }
+
+    /// Returns the number of enabled dirty qubits, ignoring the given qubits
+    #[pyo3(name = "num_dirty")]
+    fn py_num_dirty(&mut self, ignored_qubits: Vec<Qubit>) -> usize {
+        self.num_dirty(&ignored_qubits)
+    }
+
+    /// Returns the number of enabled clean qubits, ignoring the given qubits
+    #[pyo3(name = "num_clean")]
+    fn py_num_clean(&mut self, ignored_qubits: Vec<Qubit>) -> usize {
+        self.num_clean(&ignored_qubits)
+    }
+
+    /// enable using the given qubits
+    #[pyo3(name = "enable")]
+    fn py_enable(&mut self, qubits: Vec<Qubit>) {
+        self.enable(&qubits)
+    }
+
+    /// Disables using the given qubits
+    #[pyo3(name = "disable")]
+    fn py_disable(&mut self, qubits: Vec<Qubit>) {
+        self.disable(qubits.into_iter())
+    }
+
+    /// Sets state of the given qubits to dirty
+    #[pyo3(name = "set_dirty")]
+    fn py_set_dirty(&mut self, qubits: Vec<Qubit>) {
+        self.set_dirty(&qubits)
+    }
+
+    /// Sets state of the given qubits to clean
+    #[pyo3(name = "set_clean")]
+    fn py_set_clean(&mut self, qubits: Vec<Qubit>) {
+        self.set_clean(&qubits)
+    }
+
+    /// Get `num_qubits` enabled qubits, excluding `ignored_qubits`, and returning the
+    /// clean qubits first.
+    #[pyo3(name = "borrow")]
+    fn py_borrow(&mut self, num_qubits: usize, ignored_qubits: Vec<Qubit>) -> Vec<usize> {
+        self.borrow(num_qubits, &ignored_qubits)
+            .into_iter()
+            .map(|q| q.index())
+            .collect()
+    }
```

Comment:
Should we impl `ToPyObject` for the relative-index types?

Reply:
I don't feel strongly about changing anything for this particular PR (even my comment), but if we were going to, probably we ought to do the `impl IntoPyObject` in a separate PR and update other call sites, and in the mean time, this one can either stay as it is or go to `PyList`.